pkg/validator/cluster.go

```diff
@@ -312,13 +312,61 @@ func validateGeneralClusterChanges(
 	valErrors []string,
 ) []string {
 
-	if cr.Spec.AppID != prevCr.Spec.AppID {
-		appModifiedMsg := fmt.Sprintf(
-			modifiedProperty,
-			"app",
-		)
-		valErrors = append(valErrors, appModifiedMsg)
+	crApp, err := catalog.GetApp(cr)
+	if err != nil {
+		valErrors = append(valErrors, err.Error())
+	}
+
+	prevCrApp, err := catalog.GetApp(prevCr)
+	if err != nil {
+		valErrors = append(valErrors, err.Error())
 	}
+
+	if crApp != nil && prevCrApp != nil {
+
+		rbKey := cr.Name + cr.Namespace
+		rbInfo := shared.ClusterRollbackInfoMap[rbKey]
+
+		if prevCrApp.Spec.DistroID != crApp.Spec.DistroID {
+			appModifiedMsg := fmt.Sprintf(
+				invalidDistroId,
+				crApp.Spec.DistroID,
+				prevCrApp.Spec.DistroID,
+			)
+			valErrors = append(valErrors, appModifiedMsg)
+		}
+
+		if prevCrApp.Spec.Version == crApp.Spec.Version {
+			appModifiedMsg := fmt.Sprintf(
+				versionIsNotModified,
+				crApp.Spec.DistroID,
+				crApp.Spec.Version,
+			)
+			valErrors = append(valErrors, appModifiedMsg)
+		}
+
+		if !prevCrApp.Spec.Upgradable {
+			appModifiedMsg := fmt.Sprintf(
+				appNotUpgradable,
+				prevCrApp.Spec.DistroID,
+				prevCrApp.Spec.Version,
+			)
+			valErrors = append(valErrors, appModifiedMsg)
+		}
+
+		// Check, was the previous cluster configured succesfully
+		// If so, write its app identifiers as rollback info for the current one
+		if shared.ClusterIsReady(prevCr) {
+			rbInfo = &kdv1.RollbackInfo{
+				AppID:    prevCr.Spec.AppID,
+				DistroID: prevCrApp.Spec.DistroID,
+				Version:  prevCrApp.Spec.Version,
+			}
+
+			shared.ClusterRollbackInfoMap[rbKey] = rbInfo
```

Comment:
I mentioned elsewhere that an in-memory map is problematic because we'll lose it if KD restarts. I should also note that it's usually bad to do state changes as part of the validator, since even if our entire validator handler says that a change is OK it might be rejected by some other webhook and never be allowed to reach the K8s API.

So, any recording of state needs to be done in our reconciliation handlers.

For example, let's look at some of the new top level kdcluster status items we were talking about:
* a status field that reflects spec.app... let's just call it status.app here
* a status field that indicates whether an upgrade or rollback (or neither) is active... let's call it status.appChange
* a status field that indicates the app resource that we are/were upgrading from, if upgrade/rollback is active ... let's call it status.prevApp

When the kdcluster is first created, status.app will get populated with a copy of spec.app.

Now someone comes along and edits spec.app to initiate an upgrade. Our validation handler decides it's OK and lets the change go through.

The reconciliation handler sees that spec.app is different from status.app. It also sees that status.appChange is nil, so it knows an upgrade should be started. It can copy status.app to status.prevApp, copy spec.app to status.app, and set status.appChange to "upgrade".

Anyway, you get the gist. The validator just needs to greenlight the change, and then the reconciliation handler updates status and does other necessary actions like changing the statefulset and sending app notifications.

Reply:
Done. Please find the implemented changes in`pkg/controller/kubedirectorcluster/cluster.go`